features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp

```diff
@@ -0,0 +1,582 @@
+            _at.write_int(socket->id);
+            _at.write_int(blk);
+            _at.cmd_stop();
+            wait_ms(50);
+            while (ch != '@') {
+                _at.read_bytes(&ch, 1);
+            }
+            _at.write_bytes((uint8_t *)buf, blk);
+
+            _at.resp_start("+USOWR:");
+            _at.skip_param(); // skip socket id
+            sent_len = _at.read_int();
+            _at.resp_stop();
+
+            if ((sent_len >= (int) blk) &&
+                (_at.get_last_error() == NSAPI_ERROR_OK)) {
+            } else {
+                success = false;
+            }
+
+            buf += blk;
+            count -= blk;
+        }
+
+        if (success && _at.get_last_error() == NSAPI_ERROR_OK) {
+            socket->rx_avail = false;
+            return size - count;
+        }
+    }
+    _at.unlock();
+
+    return _at.get_last_error();
+}
+
+nsapi_size_or_error_t UBLOX_AT_CellularStack::socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+        void *buffer, nsapi_size_t size)
+{
+    nsapi_size_or_error_t nsapi_error_size = NSAPI_ERROR_DEVICE_ERROR;
+    bool success = true;
+    nsapi_size_t read_blk;
+    nsapi_size_t count = 0;
+    nsapi_size_t usorf_sz;
+    char ipAddress[NSAPI_IP_SIZE];
+    uint8_t ch = 0;
+    int port = 0;
+    Timer timer;
+
+    _at.lock();
+    timer.start();
+    if (socket->proto == NSAPI_UDP) {
+        while (success && (size > 0)) {
+            read_blk = UBLOX_MAX_PACKET_SIZE;
+            if (read_blk > size) {
+                read_blk = size;
+            }
+            if (socket->pending_bytes > 0) {
+                _at.cmd_start("AT+USORF=");
+                _at.write_int(socket->id);
+                _at.write_int(read_blk);
+                _at.cmd_stop();
+
+                _at.resp_start("+USORF:");
+                _at.skip_param(); // receiving socket id
+                _at.read_string(ipAddress, sizeof(ipAddress));
+                port = _at.read_int();
+                usorf_sz = _at.read_int();
+
+                // Must use what +USORF returns here as it may be less or more than we asked for
+                if (usorf_sz > socket->pending_bytes) {
+                    socket->pending_bytes = 0;
+                } else {
+                    socket->pending_bytes -= usorf_sz;
+                }
+
+                if (usorf_sz > size) {
+                    usorf_sz = size;
+                }
+                _at.read_bytes(&ch, 1);
+                _at.read_bytes((uint8_t *)buffer + count , usorf_sz);
+                _at.resp_stop();
+
+                if (usorf_sz > 0) {
+                    count += usorf_sz;
+                    size -= usorf_sz;
+                } else {
+                    // read() should not fail
+                    success = false;
+                }
+            }  else if (timer.read_ms() < SOCKET_TIMEOUT) {
+                // Wait for URCs
+                _at.process_oob();
```

Comment:
I assume this is here because modem can split a packet into multiple pieces, and it will send an URC in between the pieces... Here the call to `process_oob` is effectively a spinloop so timeout could probably be much smaller?

Reply:
Updated and reduced to 1 sec.